module/zfs/zio.c

```diff
@@ -3730,6 +3739,22 @@ zio_dva_allocate(zio_t *zio)
 	 * Fallback to normal class when an alloc class is full
 	 */
 	if (error == ENOSPC && mc != spa_normal_class(spa)) {
+		/*
+		 * When the dedup class is spilling into the normal class,
+		 * there can still be significant space available due to
+		 * deferred frees that are in-flight. We track the txg when
+		 * this occurred and back off adding new DDT entries for a
+		 * few txgs to allow the free blocks to be processed.
+		 */
```

Comment:
I believe we can be spilling here when there's only a special allocation class which is being used for dedup.

Reply:
updated to also include special class